Armel support in crossgen2

davidwrighton opened the pull request:
This change adds support for armel target compilation in crossgen2

To use the feature compile an application using the --targetos Linux and --targetarch armel flags.

Summary of changes
- To avoid bloating the compilation matrix, the #ifdefs used in the JIT that were conditional on ARM_SOFTFP behavior have been changed in to variable checks. To avoid perf penalties, this dynamic behavior is only enabled for the jit intended for compiling using crossgen2.
- Armel abi processing was integrated into the various parts of the managed crossgen2 codebase that correspond to components with special handling in the coreclr codebase
  - HFA support is disabled for Armel targets
  - Floating point registers are not used when passing arguments

With this change it becomes possible to run crossgen2 on a Windows or Linux machine and target a Linux Armel target.

davidwrighton commented:
@t-mustafin These are my changes to enable armel targetting from crossgen2. As I have no armel hardware or test environments, I have not tested the behavior of this.

@dotnet/jit-contrib This modifies the jit to be dynamically configurable for the ARM_SOFTFP and FEATURE_HFA ifdefs. I'd like some opinions on this approach, as I plan to make similar changes for the arm64 on OSX work that is ongoing, as well as possibly an attempt to compile a single arm and arm64 jit which targets both unix and Windows platforms.

BruceForstall commented on src/coreclr/src/jit/codegencommon.cpp:
```diff
@@ -3991,40 +3991,40 @@ void CodeGen::genFnPrologCalleeRegArgs(regNumber xtraReg, bool* pXtraRegClobbere
-            insCopy = ins_Copy(TYP_DOUBLE);
-            // Compute xtraReg here when we have a float argument
-            assert(xtraReg == REG_NA);
+#ifdef UNIX_AMD64_ABI
+            if (true)
+#else  // !UNIX_AMD64_ABI
+            if (GlobalJitOptions::compFeatureHfa)
+#endif // UNIX_AMD64_ABI
```
Couldn't this just be:
```
#if !defined(UNIX_AMD64_ABI)
            if (GlobalJitOptions::compFeatureHfa)
#endif // !UNIX_AMD64_ABI
```

davidwrighton replied:
Yes.

---
In reply to: [510492047](https://github.com/dotnet/runtime/pull/43706#discussion_r510492047) [](ancestors = 510492047)

BruceForstall commented on src/coreclr/src/jit/compiler.cpp:
```diff
@@ -3139,6 +3092,14 @@ void Compiler::compInitOptions(JitFlags* jitFlags)
+    opts.compUseSoftFP = !!JitConfig.JitSoftFP();
+    if (opts.compUseSoftFP)
+    {
+        GlobalJitOptions::compFeatureHfa = !opts.compUseSoftFP;
```
I'm not sure we want this as a static global. It seems like it needs to be a member of the `Compiler` object. What if we want to allow compiling both for arm and armel in the same process? Or if multiple threads try to initialize this concurrently?

Also, I don't think the config variable JitSoftFP should be the way this ABI is communicated to the JIT; it should come in via the jit flags, as a CorJitFlag.

davidwrighton replied:
The problem with compFeatureHfa is that it needs to be useable from the code in the code in LclVarDsc and fgArgTabEntry structures which don't have access to a Compiler object to query. Of course, I could pass the Compiler* as a parameter to these functions, but that didn't seem like a good approach. Given that the current model for usage of this is to only support one target architecture per process, I just went with a static. An alternative approach would be to use a threadstatic, which would allow for different compilations in the same process to have different values.

---
In reply to: [510494195](https://github.com/dotnet/runtime/pull/43706#discussion_r510494195) [](ancestors = 510494195)

BruceForstall replied:
We already stash `Compiler*` in a thread static: `JitTls::GetCompiler()`, but I'm not sure it's used in non-DEBUG code.

BruceForstall commented on src/coreclr/src/jit/compiler.cpp:
```diff
@@ -469,49 +474,6 @@ var_types Compiler::getJitGCType(BYTE gcType)
     return result;
 }
 
-#ifdef ARM_SOFTFP
```
I guess this appeared to be statically dead code?

davidwrighton replied:
Yes, this was only callable when FEATURE_HFA was true, except that ARM_SOFTFP also disables FEATURE_HFA.

---
In reply to: [510494655](https://github.com/dotnet/runtime/pull/43706#discussion_r510494655) [](ancestors = 510494655)

BruceForstall commented on src/coreclr/src/jit/compiler.h:
```diff
@@ -554,22 +552,47 @@ class LclVarDsc
 #else
-        return false;
-#endif
+        NYI("GetLvHfaElemKind");
```
These `NYI` should be `NOWAY`, since they're never going to be implemented, it not that they are "not yet implemented".

BruceForstall commented on src/coreclr/src/jit/lower.cpp:
```diff
@@ -3231,38 +3231,39 @@ void Lowering::LowerStoreLocCommon(GenTreeLclVarCommon* lclStore)
 //
 void Lowering::LowerRetStruct(GenTreeUnOp* ret)
 {
-#if defined(FEATURE_HFA) && defined(TARGET_ARM64)
```
Why remove the `#if defined(TARGET_ARM64)`? You've enabled this code for arm32. Is that correct? Maybe you can get away with it because we don't support SIMD on arm32.

davidwrighton replied:
I did not intend to make this change. Thank you for catching it.

---
In reply to: [510499387](https://github.com/dotnet/runtime/pull/43706#discussion_r510499387) [](ancestors = 510499387)

BruceForstall commented on src/coreclr/src/jit/lower.cpp:
```diff
@@ -3449,25 +3450,26 @@ void Lowering::LowerCallStruct(GenTreeCall* call)
+            // `call->HasMultiRegRetVal()` count double registers.
+            assert(comp->GetHfaCount(call) <= 2);
+#else  // !TARGET_ARM64 && !TARGET_ARM
+            NYI("Unknown architecture");
```
why change this from `unreached()` to `NYI`?

davidwrighton replied:
It turns out that part of the #if statement was never reached because of the #if which used to wrap it, and unreached() causes a compiler error now as the followon statements start producing errors. I'll change it to NOWAY from NYI per your other comment.

davidwrighton replied:
Upon reflection, this should actually probably remain as an NYI.

---
In reply to: [511061420](https://github.com/dotnet/runtime/pull/43706#discussion_r511061420) [](ancestors = 511061420)

BruceForstall reviewed (changes requested):
I'm intrigued by the ability to build a single binary to target multiple ABIs but still enable compile time specific code to avoid runtime overhead in some cases. I'm much more skeptical about something grander, such as arm/arm64 unification, compared to arm softfp, which is relatively small.

MichalStrehovsky commented on src/coreclr/src/tools/Common/TypeSystem/Common/TargetDetails.cs:
```diff
@@ -42,6 +42,10 @@ public enum TargetAbi
         /// <summary>
+        /// model for armel execution model
+        /// </summary>
+        CoreRTArmel,
```
Did you consider making this a `TargetArchitecture` instead? A separate TargetArchitecture is [how we had it](https://github.com/dotnet/corert/commit/75ccab2fb02991cd0734ad1844df060d2aa1b010#diff-b626f7f687b5636e6788091ea3d8ff43323081a1f1b0c8897322483d2247bdd3) until it got deleted because there was no use for it at that point.

We could put a helper on `TargetDetails` to help normalize those two into one for places that don't care about the distinction.

My thinking is that while `hard`/`softfp` distinction is only about the ABI, if we ever add support for `soft` - that's not just ABI anymore. I would just treat it as a very similar, but still a bit different CPU.

MichalStrehovsky replied:
(I don't have strong opinion either way, it's just we still keep `TargetAbi.Jit` and if we ever resurrect that and have to add JitArmel for parity, it feels a bit confusing.)

davidwrighton replied:
This is all very confusing, and I agree its this isn't a very stable model within the compiler. The problem is that we tend to check for architecture with a simple equality check, and there are quite a few checks (and it would be really easy to get this wrong as we don't actively test this.).

With the abi, as this change is actually specific to the abi, and we don't have many abi checks in the product, this seemed safer and less risky going forward.

davidwrighton commented:
What I meant by single arm and arm64 jits is that we would produce and arm targetting jit that handled all the various ABIs, Windows, Linux, Linux with Armel abi, and that we will generate a single arm64 jit, which targets Windows, Linux, and MacOS. I agree, merging the different architectures is a lost cause. Also, merging the X64 jit is a lot more work than the Arm one, as there are many more differences between Windows and Unix.

---
In reply to: [515161729](https://github.com/dotnet/runtime/pull/43706#issuecomment-515161729) [](ancestors = 515161729)

jashook commented on src/coreclr/src/jit/lclvars.cpp:
```diff
@@ -2821,14 +2823,15 @@ void Compiler::makeExtraStructQueries(CORINFO_CLASS_HANDLE structHandle, int lev
 void Compiler::lvaSetStructUsedAsVarArg(unsigned varNum)
 {
-#ifdef FEATURE_HFA
+    if (GlobalJitOptions::compFeatureHfa)
```
I believe this if can be removed. The existing #ifdef for FEATURE_HFA was imo redundant.

BruceForstall reviewed (approved):
LGTM

t-mustafin commented:
Sorry for a long answering.
I have run tests on armel, some tests start to fail, some tests runs too long time.

Results on [master_0e402bc](https://github.com/dotnet/runtime/commit/0e402bcd28f38a3b65720474dcd7a338cab0841d) + cherry-picked [1st_commit](https://github.com/dotnet/runtime/pull/43706/commits/f911918c4357af5c7fa6c7bbbb978decd53ae875) and [2nd_commit](https://github.com/dotnet/runtime/pull/43706/commits/8bf40517e9bd05435eb88b5074a39379b0e8ec49):
[patched.tar.gz](https://github.com/dotnet/runtime/files/5458851/patched.tar.gz)

Results on  [master_0e402bc](https://github.com/dotnet/runtime/commit/0e402bcd28f38a3b65720474dcd7a338cab0841d) for reference:
[orig.tar.gz](https://github.com/dotnet/runtime/files/5458869/orig.tar.gz)

Unfortunatelly patched.tar.gz is not a full log, cause I do not got it yet. I will restart testing with all patches from the PR and more recent master.

t-mustafin commented on src/coreclr/src/zap/zapinfo.cpp:
```diff
@@ -195,6 +195,10 @@ CORJIT_FLAGS ZapInfo::ComputeJitFlags(CORINFO_METHOD_HANDLE handle)
 #endif  // FEATURE_READYTORUN_COMPILER
 
+#ifdef ARM_SOFTFP
+    jitFlags.Set(CORJIT_FLAGS::CORJIT_FLAG_SOFTFP);
```
Got a build problem here:
```
  /home/runtime/src/coreclr/src/zap/zapinfo.cpp:199:32: error: no member named 'CORJIT_FLAG_SOFTFP' in 'CORJIT_FLAGS'; did you mean 'CORJIT_FLAG_SOFTFP_ABI'?
      jitFlags.Set(CORJIT_FLAGS::CORJIT_FLAG_SOFTFP);
                   ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
                                 CORJIT_FLAG_SOFTFP_ABI
  /home/runtime/src/coreclr/src/inc/corjitflags.h:100:9: note: 'CORJIT_FLAG_SOFTFP_ABI' declared here
          CORJIT_FLAG_SOFTFP_ABI              = 43, // On ARM should enable armel calling convention
          ^
```

davidwrighton commented:
Ah, ok, from the logs it appears that the way you are using crossgen2, you are not using the --targetarch switch to pass armel to crossgen2. And, unlike all of the other architectures, when run locally, armel isn't autodetected (as there is no platform api that will provide that bit of information.). Fortunately, I believe that is pretty simple to fix. I'll fix that, as well as the build break you noticed.

davidwrighton commented:
@t-mustafin Actually, I've just fixed the build break. I don't believe I fully understand the means by which you are running crossgen2, and the environment. As of the last PR you submitted, I believe that crossgen2 should be failing while trying to load libjitinterface_arm.so when you actually are building libjitinterface_armel.so. Did you do something like add a new enum value to the Architecture enum in System.InteropServices.RuntimeInformation that is Armel?  In any case, you should be passing the `--targetarch` switch to crossgen2 when compiling.

t-mustafin commented:
Two tests starts to fail:
```
 JIT/Methodical/Invoke/ctor/_dbgval_ctor/_dbgval_ctor.sh                                      
 JIT/Regression/JitBlue/DevDiv_462274/DevDiv_462274/DevDiv_462274.sh
```
One test starts to pass, looking at another runs seems it is unstable:
```
JIT/jit64/regress/ddb/113574/113574/113574.sh
```
Results on [master_7d15ecf](https://github.com/dotnet/runtime/commit/7d15ecf121ea848b5fa1ccb7aa0df29f1d8066dd) + cherry picked commits [1](https://github.com/dotnet/runtime/pull/43706/commits/f911918c4357af5c7fa6c7bbbb978decd53ae875) [2](https://github.com/dotnet/runtime/pull/43706/commits/8bf40517e9bd05435eb88b5074a39379b0e8ec49) [3](https://github.com/dotnet/runtime/pull/43706/commits/07e5e01299cdfaa065b4bec519c95dd15c91a849) [4](https://github.com/dotnet/runtime/pull/43706/commits/04e7fd2340585a1cd5848de1d3391c72a69535dc) [5](https://github.com/dotnet/runtime/pull/43706/commits/e3b651cdda912a3b085ee74e38eda3a825577444) [6](https://github.com/dotnet/runtime/pull/43706/commits/42211963175d2b3f808025a77b6d5fcfde8678fd)
[patched.tar.gz](https://github.com/dotnet/runtime/files/5474236/patched.tar.gz)
Results on [master_7d15ecf](https://github.com/dotnet/runtime/commit/7d15ecf121ea848b5fa1ccb7aa0df29f1d8066dd):
[results.tar.gz](https://github.com/dotnet/runtime/files/5474242/results.tar.gz)

Also I would like to run tests after cross compiling on x64 to armel, but not done it yet.

t-mustafin commented:
Results with cross generated ni files on x64.

Libs failed to cross-generate, that generates on native-run:
```
Microsoft.CodeAnalysis.CSharp.ni.dll
Microsoft.CodeAnalysis.VisualBasic.ni.dll
R2RTest/Microsoft.Build.ni.dll
```
Tests was failed to cross-generate, that native-generated successfully:
```
./JIT/Regression/JitBlue/GitHub_25027/GitHub_25027/GitHub_25027.ni.dll
```
Tests was failed after cross-generate, that pass after native-generate:
```
ilasm/PortablePdb/IlasmPortablePdbTests/IlasmPortablePdbTests.sh
ilasm/System/Runtime/CompilerServices/MethodImplOptionsTests/MethodImplOptionsTests.sh
readytorun/tests/mainv1/mainv1.sh
readytorun/tests/mainv2/mainv2.sh
```

Roughly it looks good: 11 tests fails + 13 crossgen fails compared with 112 fails on master branch.
[cross.tar.gz](https://github.com/dotnet/runtime/files/5494301/cross.tar.gz)

Does native-generated and cross-generated ni files should be equal? I got different hashes for it. 
Native:
```
# ./corerun ./crossgen2/crossgen2.dll -r:./*.dll -r:./crossgen2/*.dll --targetarch armel  -o:/opt/usr/root/hello.ni.dll  /opt/usr/root/hello.dll
Emitting R2R PE file: /opt/usr/root/hello.ni.dll
# sha1sum /opt/usr/root/hello.dll /opt/usr/root/hello.ni.dll
918acdf2054cb2e96668a01bb57610a789c0c25a  /opt/usr/root/hello.dll
9103f3dbe325b3f208d5f2b77899c5e794cad5e6  /opt/usr/root/hello.ni.dll
```
Cross:
```
$ ./corerun ./crossgen2/crossgen2.dll -r:./*.dll -r:./crossgen2/*.dll --targetarch armel  -o:/home/tmustafin/hello/hello.ni.dll  /home/tmustafin/hello/hello.dll
Emitting R2R PE file: /home/tmustafin/hello/hello.ni.dll
$ sha1sum /home/tmustafin/hello/hello.dll /home/tmustafin/hello/hello.ni.dll
918acdf2054cb2e96668a01bb57610a789c0c25a  /home/tmustafin/hello/hello.dll
9ae5debe46ac00e036fe6d68ac8d03a725a61906  /home/tmustafin/hello/hello.ni.dll
```
[hello.tar.gz](https://github.com/dotnet/runtime/files/5494446/hello.tar.gz)

cc @alpencolt

AntonLapounov commented on src/coreclr/src/jit/jit.h:
```diff
@@ -514,7 +535,7 @@ typedef ptrdiff_t ssize_t;
 #if DUMP_GC_TABLES
 #pragma message("NOTE: this non-debug build has GC ptr table dumping always enabled!")
-const bool dspGCtbls = true;
+const bool            dspGCtbls      = true;
```
I think you did not mean to insert those spaces.

ViktorHofer commented:
// Auto-generated message

https://github.com/dotnet/runtime/commit/69e114c1abf91241a0eeecf1ecceab4711b8aa62 which was merged 12/7 removed the intermediate src/coreclr/**src**/ folder. This PR needs to be updated as it touches files in that directory which causes conflicts.

To update your commits you can use this bash script: https://gist.github.com/ViktorHofer/6d24f62abdcddb518b4966ead5ef3783. Feel free to use the comment section of the gist to improve the script for others.

mangod9 commented:
@davidwrighton @t-mustafin assume we should still merge these changes? Is there anything blocking other than resolving conflicts since its a been a few months since this was opened?

t-mustafin commented:
> @davidwrighton @t-mustafin assume we should still merge these changes? Is there anything blocking other than resolving conflicts since its a been a few months since this was opened?

Yes, we need this changes. I have already appreved it, let it merge please.

mangod9 commented:
since this has been dormant for a few months, the conflicts need some manual resolving. @davidwrighton if you think they should be straightforward I can have someone shepherd this through.

clamp03 commented:
@davidwrighton Could you look into https://github.com/dotnet/runtime/issues/49347 ?

davidwrighton committed:
Make the JIT built for crossgen2 be dynamically configurable between armhf and armel calling conventions

Crossgen2 managed code changes to support Armel abi
- Adjust handling for armel target architecture command line handling to capture that armel is the abi
- Tweak field layout to disable computation of hfa status for valuetypes
- Inform JIT that armel abi is to be used
- Add armel abi value to TargetAbi enum

Checkin requested changes

Fix build on Unix platforms

Make the JIT built for crossgen2 be dynamically configurable between armhf and armel calling conventions

Crossgen2 managed code changes to support Armel abi
- Adjust handling for armel target architecture command line handling to capture that armel is the abi
- Tweak field layout to disable computation of hfa status for valuetypes
- Inform JIT that armel abi is to be used
- Add armel abi value to TargetAbi enum

Checkin requested changes

Fix build on Unix platforms

Reformat the code

Fix build break

Fix Windows build

Code review
- Use static variable, but protect from misuse with NO_WAY
- Pass info to jit via jitflag, not via jit config

tweak variable type

Address jit format issue

Fix build break

davidwrighton committed:
Fix build break

davidwrighton committed:
Use correct transition block
